Move log messages: tx enqueue to mempool, allocation to blockstorage

Sjors opened the pull request:
When running with `-debug=validation` the log is filled with:

```
2023-03-19T12:46:01Z [validation] Enqueuing TransactionAddedToMempool: txid=37c8e1ef87d75a67fbaf44e116018fbc70d9c4ef0e27e2ae56861395101f9b8e wtxid=24e2d76012f9cde451d085ead4feab136feddae6cb55d37221a81aac6c0eaf42
2023-03-19T12:46:01Z [validation] TransactionAddedToMempool: txid=37c8e1ef87d75a67fbaf44e116018fbc70d9c4ef0e27e2ae56861395101f9b8e wtxid=24e2d76012f9cde451d085ead4feab136feddae6cb55d37221a81aac6c0eaf42
```

These log entries are more appropriate for `-debug=mempool`.

This change limits `-debug=validation` to messages that are related to blocks and headers. It also downgrades `Enqueuing` messages to the `Trace` level.

It makes it easier to run for a long time without excessive log file growth. This debug category is particularly useful for keeping track of problematic headers that were seen, the timing of reorgs, see also #27276.

Since this PR touches `validationinterface.cpp` anyway, it also includes a commit to drop function names from log messages, since we now have `-logsourcelocations`.

Similarly, though less important, I moved `[validation] Pre-allocating up to position 0x… in rev….dat` messages to blockstorage. This reduces the number of `[validation]` log messages further. ~to 3 per (normal) block.~


The volume of log messages [could also have been reduced](https://github.com/bitcoin/bitcoin/pull/27277#discussion_r1141375638) by changing the log level of these messages to `trace`. Doing so would prevent a whack-a-mole game of moving log messages to different categories. However in this case I think the change in log category would be justified even if it didn't have a big impact on noise. I.e. someone running with `-debug=mempool` would expect to see the `TransactionAddedToMempool` messages and someone running with `-debug=blockstorage` would perhaps be surprised to not see the `Pre-allocating` message.

Sjors committed:
Move tx enqueue messages to mempool log

This limits -debug=validation to messages related to blocks and headers.

Sjors committed:
Don't log function names in ValidationInterface

Use -logsourcelocations instead.

Sjors commented:
Still investigating the linter issue…

```
src/validationinterface.cpp: Expected 0 argument(s) after format string but found 1 argument(s): LogPrint(category, fmt "\n", __VA_ARGS__)
```

Sjors commented on src/validationinterface.cpp:
```diff
@@ -177,18 +177,18 @@ void SyncWithValidationInterfaceQueue()
-            LOG_EVENT(fmt, local_name, __VA_ARGS__);           \
-            event();                                           \
-        });                                                    \
+#define ENQUEUE_AND_LOG_EVENT(category, event, fmt, name, ...)          \
```
Note: I don't know if there was any particular reason why the endline characters are vertically aligned. If there's no reason, then it would be easier to move them all to the left.

cc @kczyz

Sjors commented on src/flatfile.cpp:
```diff
@@ -66,7 +66,7 @@ size_t FlatFileSeq::Allocate(const FlatFilePos& pos, size_t add_size, bool& out_
             FILE *file = Open(pos);
             if (file) {
-                LogPrint(BCLog::VALIDATION, "Pre-allocating up to position 0x%x in %s%05u.dat\n", new_size, m_prefix, pos.nFile);
+                LogPrint(BCLog::BLOCKSTORE, "Pre-allocating up to position 0x%x in %s%05u.dat\n", new_size, m_prefix, pos.nFile);
```
I think blockstorage is correct place for this in any case. E.g. it also logs `Leaving block file`. I don't know if it makes sense to reduce the log level.

Sjors replied:
Looking at the history of this category, it was introduced as part of a larger change in #23235 to reduce the default log noise. There was no discussion about this particular line, and my guess is that it was overlooked. cc @ajtowns

Sjors replied:
(added a note to the PR description)

Sjors replied:
I'll look into switching to the `LogPrintLevel` function.

Sjors replied:
Done

ajtowns replied:
> Looking at the history of the `BLOCKSTORE` category, it was introduced as part of a larger change in #23235 to reduce the default log noise. There was no discussion about this particular line, and my guess is that it was overlooked. cc @ajtowns

That sounds right

0xB10C commented:
> CI network error https://cirrus-ci.com/task/4864754720702464?logs=ci#L61:
> 
> `TimeoutError: [Errno 110] Connection timed out`

Note: The failed connections seems to be to the launchpad.net PPA with a newer bpfcc. Not using this newer PPA to install an up-to-date `bpfcc` causes warnings printed to stderr which means `interface_usdt_coinselection.py` is treated as failed. (The up-coming Ubuntu lunar seems to include an up-to-date bpfcc).

ajtowns commented:
Approach ACK

Sjors commented:
Rebased after #27632.

Sjors commented on src/flatfile.cpp:
```diff
@@ -66,7 +66,7 @@ size_t FlatFileSeq::Allocate(const FlatFilePos& pos, size_t add_size, bool& out_
             FILE *file = Open(pos);
             if (file) {
-                LogPrint(BCLog::VALIDATION, "Pre-allocating up to position 0x%x in %s%05u.dat\n", new_size, m_prefix, pos.nFile);
+                LogPrintLevel(BCLog::BLOCKSTORAGE, BCLog::Level::Info, "Pre-allocating up to position 0x%x in %s%05u.dat\n", new_size, m_prefix, pos.nFile);
```
It's more complicated. `WillLogCategoryLevel` logs `>= BCLog::Level::Warning` for any category. For `Info` it first checks if the category is selected, which is only the case if the user set `-debug=blockstorage`.

Only if the category is set, it looks at category-specific log levels (`-loglevel`), and defaults to `DEFAULT_LOG_LEVEL`, which is currently `Debug` (but I vaguely recall the idea was the raise that to `Info`).

So this should _not_ cause logging by default. I could still set the level to `Debug` but perhaps that's better reserver for even lower level block storage stuff, if we ever add that. And indeed it didn't when I briefly tested it .

ajtowns replied:
We have `Trace` available for "even lower level" stuff (currently used by the wallet for sql logging). The only `Info` logging we currently do with a category is for libevent `EVENT_LOG_MSG`, which I don't think is even used in the parts of libevent we actually use. I don't think this change makes sense, but that's probably mostly thinking that the logging api we have at the moment is confusing... (cf #28318)

Sjors replied:
I changed it to `Debug`

Sjors committed:
Move pre-allocating message to blockstorage log

ajtowns commented on src/validationinterface.cpp:
```diff
@@ -177,18 +177,17 @@ void SyncWithValidationInterfaceQueue()
-#define LOG_EVENT(fmt, ...) \
-    LogPrint(BCLog::VALIDATION, fmt "\n", __VA_ARGS__)
+#define LOG_EVENT(category, level, fmt, ...) \
+    LogPrintLevel(category, level, fmt "\n", __VA_ARGS__)
```
There's not really any need for a separate `LOG_EVENT` macro anymore -- all it saves now is adding a newline.

Sjors replied:
Got rid of it.

ajtowns commented on src/validationinterface.cpp:
```diff
@@ -198,7 +197,7 @@ void CMainSignals::UpdatedBlockTip(const CBlockIndex *pindexNew, const CBlockInd
         m_internals->Iterate([&](CValidationInterface& callbacks) { callbacks.UpdatedBlockTip(pindexNew, pindexFork, fInitialDownload); });
     };
-    ENQUEUE_AND_LOG_EVENT(event, "%s: new block hash=%s fork block hash=%s (in IBD=%s)", __func__,
+    ENQUEUE_AND_LOG_EVENT(BCLog::VALIDATION, BCLog::Level::Info, event, "new block hash=%s fork block hash=%s (in IBD=%s)",
```
Also seems pointless to upgrade these to `Info`... If anything, these seem more appropriate to downgrade to `Trace` logging (the "Enqueuing" particularly).

Sjors replied:
I'll switch these to `Debug` and downgrade the `Enqueuing` wrapper to trace.

ajtowns commented:
> This reduces the number of `[validation]` log messages to 3 per (normal) block.

I'm seeing 6 validation logs via validationinterface.cpp: `NewPoWValidBlock`, `BlockChecked`, `BlockConnected` (enqueued and executed), `UpdatedBlockTip` (enqueued and executed). I think that's with default settings.

There's also "Saw new header", "Saw new cmpctblock header" and "UpdateTip", but those aren't `[validation]` log messages (they're not categorised and don't require `-debug=validation`), but if you're only seeing those, you already weren't seeing the "Pre-allocating..." messages...

ajtowns commented on src/validationinterface.cpp:
```diff
@@ -177,18 +177,17 @@ void SyncWithValidationInterfaceQueue()
-            LOG_EVENT(fmt, local_name, __VA_ARGS__);           \
-            event();                                           \
-        });                                                    \
+#define ENQUEUE_AND_LOG_EVENT(category, level, event, fmt, ...)    \
```
No real need to include the `level` as a param, I think. Hardcoding it would make things a bit less noisy.

Sjors replied:
This PR always sets `level` to `Debug`, so we don't need it here. But then any later tweak in log levels would have to edit this macro again.

Sjors commented:
Rebased and addressed feedback.

> I'm seeing 6 validation logs via `validationinterface.cpp:`

I don't recall how I came up with `3`. It's indeed more, sometimes many more (when receiving the same header from multiple peers?). I may have gotten confused with #27826. Updated the description.

Sjors commented:
Oops... 

```
Error "tinyformat: Not enough conversion specifiers in format string" while formatting log message: block hash=%s
```

Will push fix.

Sjors commented:
Dropping the function name (by default) made some of these messages ambiguous so I clarified them a bit in f6364e332db0eb19280cd3d8f10492439cc34948.

Sjors commented on src/validationinterface.cpp:
```diff
@@ -246,17 +242,17 @@ void CMainSignals::ChainStateFlushed(const CBlockLocator &locator) {
         m_internals->Iterate([&](CValidationInterface& callbacks) { callbacks.ChainStateFlushed(locator); });
     };
-    ENQUEUE_AND_LOG_EVENT(event, "%s: block hash=%s", __func__,
+    ENQUEUE_AND_LOG_EVENT(BCLog::VALIDATION, BCLog::Level::Debug, event, "chainstate flushed block hash=%s",
```
As an aside, this gets called 5-10 times for some blocks, zero times for most blocks.

Sjors commented:
Rebased after #28368.

Sjors commented:
Rebased and switched `LogPrintLevel(...Debug)` to `LogDebug(...)`

Sjors committed:
Update validationinterface.cpp format strings

Sjors committed:
refactor: add category and level args to ENQUEUE_AND_LOG_EVENT

The Enqueuing wrapper uses level Trace.

Sjors commented on src/validationinterface.cpp:
```diff
@@ -179,19 +179,16 @@ void SyncWithValidationInterfaceQueue()
+#define ENQUEUE_AND_LOG_EVENT(category, level, event, fmt, name, ...)                                 \
+    do {                                                                                              \
+        auto local_name = (name);                                                                     \
+        LogTrace(category, "Enqueuing " fmt "\n", local_name, __VA_ARGS__); \
```
Indeed this should be `LogDebug()` in order to not change behavior.

The downgrade to trace should be in a different commit. Will look into this again.

>  This makes intermittent test failures harder to debug, since trace is disabled.

I'll look into what level to use. Lots of things have changed over the past few rebases.

Sjors commented:
Let's mark this up for grabs. Though I might eventually grab it.
